ee/clickhouse/queries/clickhouse_retention.py

```diff
@@ -58,62 +50,103 @@ def calculate_retention(self, filter: Filter, team: Team, total_intervals: int)
+                        """
+            )
+            .update_params(
+                start_date=date_from.strftime(
                     "%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")
                 ),
-                "end_date": date_to.strftime(
+                end_date=date_to.strftime("%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")),
+                team_id=team.pk,
+            )
+            .c(FROM=f"events e JOIN ({person_q}) pdi on e.distinct_id = pdi.distinct_id")
+            .update_params(**person_q.params),
+            lambda q: target_query and q.c(WHERE=target_query).update_params(**target_params),
+            lambda q: prop_filters and q.c(WHERE=prop_filters).update_params(**prop_filter_params),
+        )
+
+        reference_event_query = compose(
+            lambda: SqlQuery(
+                SELECT="""
+                        DISTINCT pdi.person_id as person_id,
+                        {trunc_func}(e.timestamp) as event_date
+                        """.format(
+                    trunc_func=trunc_func
+                )
+            )
+            .c(
+                WHERE="""
+                        toDateTime(e.timestamp) >= toDateTime(%(start_date)s)
+                        AND toDateTime(e.timestamp) <= toDateTime(%(end_date)s)
+                        AND e.team_id = %(team_id)s
+                        """
+            )
+            .update_params(
+                start_date=date_from.strftime(
                     "%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")
                 ),
-                **prop_filter_params,
-                **target_params,
-                "period": period,
-            },
+                end_date=date_to.strftime("%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")),
+                team_id=team.pk,
+            )
+            .c(FROM=f"events e JOIN ({person_q}) pdi on e.distinct_id = pdi.distinct_id")
+            .update_params(**person_q.params),
+            lambda q: target_query and q.c(WHERE=target_query).update_params(**target_params),
+            lambda q: prop_filters and q.c(WHERE=prop_filters).update_params(**prop_filter_params),
         )
 
+        final_query = SqlQuery(
+            SELECT="""
+                datediff(%(period)s, {trunc_func}(toDateTime(%(start_date)s)), reference_event.event_date) as period_to_event_days,
+                datediff(%(period)s, reference_event.event_date, {trunc_func}(toDateTime(event_date))) as period_between_events_days,
+                COUNT(DISTINCT event.person_id) count
+            """.format(
+                trunc_func=trunc_func
+            ),
+            FROM=f"({event_query}) event JOIN ({reference_event_query}) reference_event ON (event.person_id = reference_event.person_id)",
+            WHERE="{trunc_func}(event.event_date) >= {trunc_func}(reference_event.event_date)".format(
```

Comment:
Note: trunct_func is scattered across many different places here - I'm wondering if we could use composition here to make it cleaner.